File: Medium/Find the number of islands/find-the-number-of-islands.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while (T-- > 0) {
            String[] s = br.readLine().trim().split(" ");
            int n = Integer.parseInt(s[0]);
            int m = Integer.parseInt(s[1]);
            char[][] grid = new char[n][m];
            for (int i = 0; i < n; i++) {
                String[] S = br.readLine().trim().split(" ");
                for (int j = 0; j < m; j++) {
                    grid[i][j] = S[j].charAt(0);
                }
            }
            Solution obj = new Solution();
            int ans = obj.numIslands(grid);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends

class pair
{
    int row;
    int col;
    pair(int row,int col)
    {
        this.row=row;
        this.col=col;
    }
}
class Solution {
    public boolean isValid(int i,int j,int m,int n)
    {
        return i>=0 && i<m && j>=0 && j<n;
    }
    public void bfs(int r,int c,char grid[][],boolean visited[][])
    {
        visited[r][c]=true;
        Queue<pair> q=new LinkedList<>();
        q.offer(new pair(r,c));
        while(!q.isEmpty())
        {
            pair top=q.poll();
            int row=top.row;
            int col=top.col;
            for(int i=-1;i<=1;i++)
            {
                for(int j=-1;j<=1;j++)
                {
                    int x=row+i;
                    int y=col+j;
                    if(isValid(x,y,grid.length,grid[0].length))
                    {
                        if(grid[x][y]=='1' && !visited[x][y])
                        {
                            visited[x][y]=true;
                            q.offer(new pair(x,y));
                        }
                    }
                }
            }
        }
    }
    // Function to find the number of islands.
    public int numIslands(char[][] grid) {
        // Code here
        int m=grid.length,n=grid[0].length;
        int i,j;
        boolean visited[][]=new boolean[m][n];
        for(i=0;i<m;i++)
        Arrays.fill(visited[i],false);
        int count=0;
        for(i=0;i<m;i++)
        {
            for(j=0;j<n;j++)
            {
                if(grid[i][j]=='1' && !visited[i][j]){
                count++;
                bfs(i,j,grid,visited);
                }
            }
        }
        return count;
    }
}

File: Medium/Find the number of islands/find-the-number-of-islands.py
#User function Template for python3

import sys
from collections import deque

sys.setrecursionlimit(10**8)

class Solution:
    def bfs(self, i, j, v, grid):
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1), (-1, 1), (1, -1), (1, 1), (-1, -1)]
        v[i][j] = 1
        q = deque([(i, j)])
        while q:
            a, b = q.popleft()
            for dx, dy in directions:
                x, y = a + dx, b + dy
                if 0 <= x < len(grid) and 0 <= y < len(grid[0]) and grid[x][y] == 1 and not v[x][y]:
                    v[x][y] = 1
                    q.append((x, y))

    def numIslands(self, grid):
        n = len(grid)
        m = len(grid[0])
        cnt = 0
        v = [[0 for _ in range(m)] for _ in range(n)]
        for i in range(n):
            for j in range(m):
                if not v[i][j] and grid[i][j] == 1:
                    cnt += 1
                    self.bfs(i, j, v, grid)
        return cnt


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__=="__main__":
    for _ in range(int(input())):
        n,m=map(int,input().strip().split())
        grid=[]
        for i in range(n):
            grid.append([int(i) for i in input().strip().split()])
        obj=Solution()
        print(obj.numIslands(grid))
# } Driver Code Ends